File: archive/c/c-sharp/MinimumSpanningTree.cs
if (args is not [var input] || !TryParseMatrix(input.AsSpan(), out var matrix, out int n))
    return ExitWithUsage();

int weight = MinimumSpanningTreeWeight(matrix, n);
if (weight < 0)
    return ExitWithUsage();

Console.WriteLine(weight);
return 0;

static int MinimumSpanningTreeWeight(List<int> matrix, int n)
{
    var inMst = new bool[n];
    var minEdge = new int[n];

    Array.Fill(minEdge, int.MaxValue);

    minEdge[0] = 0;

    int total = 0;

    for (int added = 0; added < n; added++)
    {
        int bestWeight = int.MaxValue;
        int u = -1;

        for (int i = 0; i < n; i++)
        {
            if (!inMst[i] && minEdge[i] < bestWeight)
            {
                bestWeight = minEdge[i];
                u = i;
            }
        }

        if (u < 0)
            return -1;

        inMst[u] = true;
        total += bestWeight;

        int row = u * n;

        for (int v = 0; v < n; v++)
        {
            int w = matrix[row + v];

            if (w != 0 && !inMst[v] && w < minEdge[v])
                minEdge[v] = w;
        }
    }

    return total;
}

static bool TryParseMatrix(ReadOnlySpan<char> view, out List<int> numbers, out int dimension)
{
    numbers = null!;
    dimension = 0;

    if (view.IsWhiteSpace())
        return false;

    int expected = view.Count(',') + 1;
    if (expected < 4) return false;

    var list = new List<int>(expected);

    while (!view.IsEmpty)
    {
        int i = view.IndexOf(',');
        var token = i >= 0 ? view[..i] : view;

        view = i >= 0 ? view[(i + 1)..] : [];

        if (!int.TryParse(token, out int v))
            return false;

        list.Add(v);
    }

    int d = (int)Math.Sqrt(list.Count);
    if (d * d != list.Count)
        return false;

    numbers = list;
    dimension = d;
    return true;
}

static int ExitWithUsage()
{
    Console.Error.WriteLine("""Usage: please provide a comma-separated list of integers""");
    return 1;
}

File: archive/v/visual-basic/minimum-spanning-tree.vb
Public Module MinimumSpanningTree

    Private Sub ShowUsage()
        Console.Error.WriteLine(
            "Usage: please provide a comma-separated list of integers"
        )
        Environment.Exit(1)
    End Sub


    Private Function ParseMatrix(input As String) As List(Of Integer)

        If String.IsNullOrWhiteSpace(input) Then
            ShowUsage()
        End If

        Dim result As New List(Of Integer)

        For Each token In input.Split(","c, StringSplitOptions.RemoveEmptyEntries Or StringSplitOptions.TrimEntries)

            Dim value As Integer
            If Not Integer.TryParse(token, value) Then
                ShowUsage()
            End If

            result.Add(value)

        Next

        If result.Count < 4 Then
            ShowUsage()
        End If

        Dim dimension As Integer = CInt(Math.Sqrt(result.Count))

        If dimension * dimension <> result.Count Then
            ShowUsage()
        End If

        Return result

    End Function


    Private Function MinimumSpanningTreeWeight(matrix As List(Of Integer), dimension As Integer) As Integer

        Dim inMST(dimension - 1) As Boolean
        Dim minEdge(dimension - 1) As Integer

        For i As Integer = 0 To dimension - 1
            minEdge(i) = Integer.MaxValue
        Next

        minEdge(0) = 0
        Dim totalWeight As Integer = 0

        For i As Integer = 0 To dimension - 1

            Dim minValue As Integer = Integer.MaxValue
            Dim node As Integer = -1

            For j As Integer = 0 To dimension - 1
                If Not inMST(j) AndAlso minEdge(j) < minValue Then
                    minValue = minEdge(j)
                    node = j
                End If
            Next

            If node = -1 Then
                ShowUsage()
            End If

            inMST(node) = True
            totalWeight += minValue

            For adj As Integer = 0 To dimension - 1

                Dim weight As Integer = matrix(node * dimension + adj)

                If weight <> 0 AndAlso Not inMST(adj) AndAlso weight < minEdge(adj) Then
                    minEdge(adj) = weight
                End If

            Next

        Next

        Return totalWeight

    End Function


    Public Function Main(args As String()) As Integer

        If args.Length <> 1 Then
            ShowUsage()
        End If

        Dim matrix = ParseMatrix(args(0))
        Dim dimension As Integer = CInt(Math.Sqrt(matrix.Count))

        Dim result = MinimumSpanningTreeWeight(matrix, dimension)

        Console.WriteLine(result)

        Return 0

    End Function

End Module